File: usdQt/primIdTable.cpp
//
// Copyright 2016 Pixar
//
// Licensed under the Apache License, Version 2.0 (the "Apache License")
// with the following modification; you may not use this file except in
// compliance with the Apache License and the following modification to it:
// Section 6. Trademarks. is deleted and replaced with:
//
// 6. Trademarks. This License does not grant permission to use the trade
//    names, trademarks, service marks, or product names of the Licensor
//    and its affiliates, except as required to comply with Section 4(c) of
//    the License and to reproduce the content of the NOTICE file.
//
// You may obtain a copy of the Apache License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License with the above modification is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the Apache License for the specific
// language governing permissions and limitations under the Apache License.
////
// Copyright 2016 Pixar
//
// Licensed under the Apache License, Version 2.0 (the "Apache License")
// with the following modification; you may not use this file except in
// compliance with the Apache License and the following modification to it:
// Section 6. Trademarks. is deleted and replaced with:
//
// 6. Trademarks. This License does not grant permission to use the trade
//    names, trademarks, service marks, or product names of the Licensor
//    and its affiliates, except as required to comply with Section 4(c) of
//    the License and to reproduce the content of the NOTICE file.
//
// You may obtain a copy of the Apache License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the Apache License with the above modification is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the Apache License for the specific
// language governing permissions and limitations under the Apache License.
//

#include <iostream>
#include "pxr/base/tf/stringUtils.h"

#include "primIdTable.h"
#include "debug.h"

UsdQt_PrimIdTable::UsdQt_PrimIdTable(const UsdStagePtr& stage,
                                     const UsdPrim& root,
                                     const Usd_PrimFlagsPredicate& predicate,
                                     UsdQt_InternalId maxId)
    : _stage(stage),
      _predicate(predicate),
      _root(root.GetPath()),
      _nextAvailableId(1),
      _maxId(maxId) {
    //    TfDebug::Enable(USDQT_DEBUG_PRIMIDTABLE);

    TF_VERIFY(_maxId > _nextAvailableId);
    _RegisterPrim(root);
}

void UsdQt_PrimIdTable::_RegisterPrim(const UsdPrim& prim) {
    SdfPath path = prim.GetPath();
    TF_VERIFY(_nextAvailableId < _maxId);
    if (_pathToId.count(path) < 1) {
        _pathToId[path] = _nextAvailableId;
        UsdQt_PrimIdTable::_ItemInfo info;
        info.path = path;
        for (auto child : prim.GetFilteredChildren(_predicate))
            info.children.push_back(child.GetPath());

        _idToItem[_nextAvailableId] = info;
        _nextAvailableId++;
    }
}

bool UsdQt_PrimIdTable::IsRoot(UsdQt_InternalId id) const {
    auto itemIterator = _idToItem.find(id);
    return itemIterator != _idToItem.end() &&
           itemIterator->second.path == _root;
}

bool UsdQt_PrimIdTable::ContainsPath(const SdfPath& path) const {
    return _pathToId.count(path) > 0;
}

UsdQt_InternalId UsdQt_PrimIdTable::GetIdFromPath(const SdfPath& path) const {
    auto idIterator = _pathToId.find(path);
    if (idIterator == _pathToId.end()) {
        TF_CODING_ERROR("Cannot find '%s' in PrimIdTable", path.GetText());
        return 0;
    }
    return idIterator->second;
}

SdfPath UsdQt_PrimIdTable::GetPathFromId(UsdQt_InternalId id) const {
    auto itemIterator = _idToItem.find(id);
    if (itemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find '%u' in PrimIdTable", id);
        return SdfPath();
    }
    return itemIterator->second.path;
}

size_t UsdQt_PrimIdTable::GetChildCount(UsdQt_InternalId id) const {
    auto itemIterator = _idToItem.find(id);
    if (itemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find '%u' in PrimIdTable", id);
        return 0;
    }
    return itemIterator->second.children.size();
}

SdfPath UsdQt_PrimIdTable::GetChildPath(UsdQt_InternalId id,
                                        size_t index) const {
    auto itemIterator = _idToItem.find(id);
    if (itemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find '%u' in PrimIdTable", id);
        return SdfPath();
    }

    if (index >= itemIterator->second.children.size()) {
        TF_CODING_ERROR(
            "Index '%zu' exceeds number of children of '%s' in PrimIdTable",
            index, itemIterator->second.path.GetText());
        return SdfPath();
    }
    return itemIterator->second.children[index];
}

UsdQt_InternalId UsdQt_PrimIdTable::GetParentId(UsdQt_InternalId id) const {
    auto itemIterator = _idToItem.find(id);
    if (itemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find '%u' in PrimIdTable", id);
        return 0;
    }
    const SdfPath& path = itemIterator->second.path;
    if (path == _root) {
        return 0;
    }
    SdfPath parent = path.GetParentPath();
    auto parentIdIterator = _pathToId.find(parent);
    if (parentIdIterator == _pathToId.end()) {
        TF_CODING_ERROR("Cannot find parent '%s' of '%s' in PrimIdTable",
                        parent.GetText(), path.GetText());
        return 0;
    }

    return parentIdIterator->second;
}

size_t UsdQt_PrimIdTable::GetRow(UsdQt_InternalId id) const {
    auto itemIterator = _idToItem.find(id);
    if (itemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find '%u' in PrimIdTable", id);
        return 0;
    }

    UsdQt_InternalId parentId = GetParentId(id);
    SdfPath path = GetPathFromId(id);
    if (path == _root) {
        return 0;
    }

    auto parentItemIterator = _idToItem.find(parentId);
    if (parentItemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find parent '%u' of '%u' in PrimIdTable",
                        parentId, id);
        return 0;
    }

    // Run a search to find the index.
    auto rowIterator =
        std::find(parentItemIterator->second.children.begin(),
                  parentItemIterator->second.children.end(), path);

    return rowIterator - parentItemIterator->second.children.begin();
}

bool UsdQt_PrimIdTable::RegisterChild(UsdQt_InternalId id, size_t index) {
    auto itemIterator = _idToItem.find(id);
    if (itemIterator == _idToItem.end()) {
        TF_CODING_ERROR("Cannot find '%u' in PrimIdTable", id);
        return false;
    }
    if (index >= itemIterator->second.children.size()) {
        TF_CODING_ERROR(
            "Index '%zu' exceeds number of children of '%s' in PrimIdTable",
            index, itemIterator->second.path.GetText());
        return false;
    }

    const SdfPath& path = itemIterator->second.children[index];

    if (_pathToId.count(path)) {
        // already registered.
        return true;
    }
    UsdPrim prim = _stage->GetPrimAtPath(path);
    if (not prim) {
        TF_CODING_ERROR("Expected child (%zu) of id (%u) has expired.", index,
                        id);
        return false;
    }
    if (_nextAvailableId >= _maxId) {
        TF_WARN("out of indices.");
        return false;
    }

    _RegisterPrim(prim);
    return true;
}

void UsdQt_PrimIdTable::_DeleteSubtree(const SdfPath& path) {
    if (_pathToId.count(path) > 0) {
        TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
            .Msg("Deleting instantiated path: '%s'\n", path.GetText());
        UsdQt_InternalId id = _pathToId[path];
        _pathToId.erase(path);
        for (auto childPath : _idToItem[id].children) {
            _DeleteSubtree(childPath);
        }
        _idToItem.erase(id);
    } else {
        TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE).Msg(
            "Skipping deletion of uninstantiated path: '%s'\n", path.GetText());
    }
}

void UsdQt_PrimIdTable::_InvalidateSubtree(const SdfPath& path) {
    if (_pathToId.count(path) > 0) {
        UsdQt_InternalId id = _pathToId[path];
        UsdPrim prim = _stage->GetPrimAtPath(path);
        if (prim.IsValid() and _predicate(prim)) {
            TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
                .Msg("Keeping '%s' during invalidation.\n", path.GetText());
            for (auto childPath : _idToItem[id].children) {
                _InvalidateSubtree(childPath);
            }
            TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
                .Msg("Original size: %zu children.\n",
                     _idToItem[id].children.size());
            _idToItem[id].children.clear();
            for (auto child : prim.GetFilteredChildren(_predicate)) {
                _idToItem[id].children.push_back(child.GetPath());
            }
            TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE).Msg(
                "New size: %zu children.\n", _idToItem[id].children.size());
        } else {
            TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
                .Msg("Rejecting '%s' during invalidation.\n", path.GetText());
            _DeleteSubtree(path);
        }
    } else {
        TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
            .Msg("Skipping invalidation of uninstantiated path '%s'\n",
                 path.GetText());
    }
}

void UsdQt_PrimIdTable::ResyncSubtrees(const std::vector<SdfPath>& paths) {
    /// TODO: Is there a way to optimize the paths that need to be traversed
    /// if there are redundancies? (say, /World/foo and /World/foo/bar)
    std::vector<SdfPath> sortedPaths(paths);
    std::sort(sortedPaths.begin(), sortedPaths.end());

    /// Uniquify the list of parents.
    std::set<SdfPath> uniqueParents;
    for (auto path : sortedPaths) uniqueParents.insert(path.GetParentPath());

    std::set<SdfPath> outOfSyncPaths;

    // Update the list of children per unique parent.
    for (auto parentPath : uniqueParents) {
        TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
            .Msg("Updating children of parent: '%s'\n", parentPath.GetText());

        UsdQt_InternalId parentId = GetIdFromPath(parentPath);
        std::vector<SdfPath> sortedOriginalChildren =
            _idToItem[parentId].children;
        std::sort(sortedOriginalChildren.begin(), sortedOriginalChildren.end());

        std::vector<SdfPath> newChildren;

        // Look through the new children to find any paths aren't in the
        // original
        // children and not in sorted paths.  These have gotten out of sync.
        UsdPrim parentPrim = _stage->GetPrimAtPath(parentPath);
        for (auto child : parentPrim.GetFilteredChildren(_predicate)) {
            bool inOriginalChildren = std::binary_search(
                sortedOriginalChildren.begin(), sortedOriginalChildren.end(),
                child.GetPath());
            bool inResyncPaths = std::binary_search(
                sortedPaths.begin(), sortedPaths.end(), child.GetPath());

            if (inOriginalChildren or inResyncPaths) {
                TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
                    .Msg("Keeping child: '%s'\n", child.GetPath().GetText());
                newChildren.push_back(child.GetPath());
            } else {
                TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE).Msg(
                    "Out of sync new child: '%s'\n", child.GetPath().GetText());
                outOfSyncPaths.insert(child.GetPath());
            }
        }

        std::vector<SdfPath> sortedNewChildren(newChildren);
        std::sort(sortedNewChildren.begin(), sortedNewChildren.end());
        // Look through the original children to find any paths are missing
        // and not in sorted paths.  These have gotten out of sync, likely
        // because ResyncSubtrees has been called with an incomplete list.
        // This isn't strictly necessary other than for error checking.
        for (auto childPath : sortedOriginalChildren) {
            bool inNewChildren = std::binary_search(
                sortedNewChildren.begin(), sortedNewChildren.end(), childPath);
            bool inResyncPaths = std::binary_search(
                sortedPaths.begin(), sortedPaths.end(), childPath);
            if (not inNewChildren and not inResyncPaths) {
                TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE).Msg(
                    "Out of sync original child: '%s'\n", childPath.GetText());
                outOfSyncPaths.insert(childPath);
            }
        }
        // Assign the new children vector to the parent.
        _idToItem[parentId].children = newChildren;
        TF_DEBUG(USDQT_DEBUG_PRIMIDTABLE)
            .Msg("Total children count: '%zu'\n",
                 _idToItem[parentId].children.size());
    }

    if (outOfSyncPaths.size() > 0) {
        TF_CODING_ERROR("Indices may have been lost during index resync.");
    }
    for (auto path : sortedPaths) _InvalidateSubtree(path);
}

void UsdQt_PrimIdTable::PrintFullIndex() {
    std::cout << "Root: " << _root << std::endl;
    for (auto it : _idToItem) {
        std::string pathMapInfo;
        if (_pathToId.count(it.second.path)) {
            if (_pathToId[it.second.path] == it.first) {
                pathMapInfo = "correct";
            } else {
                pathMapInfo = TfStringPrintf("out of sync path map entry: %u",
                                             _pathToId[it.second.path]);
            }
        } else {
            pathMapInfo = "Missing Path Map Entry.";
        }
        std::cout << "[id]: " << it.first << " [path]: " << it.second.path
                  << " [path map]: " << pathMapInfo
                  << " [child count]: " << it.second.children.size()
                  << std::endl;
    }
    for (auto it : _pathToId) {
        if (_idToItem.count(it.second) < 1) {
            std::cout << "Dangling path: " << it.first << std::endl;
        }
    }
}

void UsdQt_PrimIdTable::PrintSubtreeIndex(const SdfPath& path) {
    if (_pathToId.count(path) < 1) {
        std::cout << "(uninstantiated) " << path.GetString() << std::endl;
    } else {
        UsdQt_InternalId id = GetIdFromPath(path);
        std::cout << path << " id: " << id << " row: " << GetRow(id)
                  << std::endl;
        for (auto child : _idToItem[id].children) PrintSubtreeIndex(child);
    }
}

File: usdQt/primIdTable.py
#
# Copyright 2016 Pixar
#
# Licensed under the Apache License, Version 2.0 (the "Apache License")
# with the following modification you may not use this file except in
# compliance with the Apache License and the following modification to it:
# Section 6. Trademarks. is deleted and replaced with:
#
# 6. Trademarks. This License does not grant permission to use the trade
#    names, trademarks, service marks, or product names of the Licensor
#    and its affiliates, except as required to comply with Section 4(c) of
#    the License and to reproduce the content of the NOTICE file.
#
# You may obtain a copy of the Apache License at
#
#     http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the Apache License with the above modification is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the Apache License for the specific
# language governing permissions and limitations under the Apache License.
#

from __future__ import absolute_import, print_function
import sys
import logging
import bisect
from pxr import Usd, Sdf, Tf

from typing import Dict, List, NamedTuple, Set

_ItemInfo = NamedTuple('_ItemInfo', [
    ('path', Sdf.Path),   # The path represented by the item id.
    ('children', List[Sdf.Path])
])

_InternalId = int

_logger = logging.getLogger(__name__)


def binarySearch(a, x):
    i = bisect.bisect_left(a, x)
    return i < len(a)


class _PrimIdTable(object):
    def __init__(self, stage, root, predicate=Usd.PrimDefaultPredicate,
                 maxId=sys.maxint):
        """
        Parameters
        ----------
        stage : Usd.Stage
        root : Usd.Prim
        predicate: Usd.PrimDefaultPredicate
        maxId : _InternalId 

        Returns
        -------
        None
        """
        self._stage = stage
        self._predicate = predicate
        self._root = root.GetPath()
        self._nextAvailableId = 1
        self._maxId = maxId
        self._pathToId = {}  # type: Dict[Sdf.Path, _InternalId]
        self._idToItem = {}  # type: Dict[_InternalId, _ItemInfo]
        #    TfDebug.Enable(USDQT_DEBUG_PRIMIDTABLE)

        assert self._maxId > self._nextAvailableId
        self._RegisterPrim(root)

    def _RegisterPrim(self, prim):
        """
        Parameters
        ----------
        prim : Usd.Prim

        Returns
        -------
        None
        """
        path = prim.GetPath()
        assert self._nextAvailableId < self._maxId
        if path not in self._pathToId:
            self._pathToId[path] = self._nextAvailableId
            info = _ItemInfo(path, [])
            for child in prim.GetFilteredChildren(self._predicate):
                info.children.append(child.GetPath())

            self._idToItem[self._nextAvailableId] = info
            self._nextAvailableId += 1

    def GetPredicate(self):
        """
        Returns
        -------
        Usd.PrimFlagsPredicate
        """
        return self._predicate

    def IsRoot(self, id):
        """
        Parameters
        ----------
        id : _InternalId

        Returns
        -------
        bool
        """
        return self._idToItem.get(id) == self._root

    def GetRootPath(self):
        """
        Returns
        -------
        Sdf.Path
        """
        return self._root

    def ContainsPath(self, path):
        """
        Parameters
        ----------
        path : Sdf.Path

        Returns
        -------
        bool
        """
        return path in self._pathToId

    def ContainsId(self, id):
        """
        Parameters
        ----------
        id : _InternalId

        Returns
        -------
        bool
        """
        return id in self._idToItem

    def GetIdFromPath(self, path):
        """
        Parameters
        ----------
        path : Sdf.Path

        Returns
        -------
        _InternalId
        """
        id = self._pathToId.get(path)
        if id is None:
            raise Tf.ErrorException("Cannot find '%s' in PrimIdTable" % path)
        return id

    def GetPathFromId(self, id):
        """
        Parameters
        ----------
        id : _InternalId

        Returns
        -------
        Sdf.Path
        """
        item = self._idToItem.get(id)
        if item is None:
            raise Tf.ErrorException("Cannot find '%d' in PrimIdTable" % id)
        return item.path

    def GetChildCount(self, id):
        """
        Parameters
        ----------
        id : _InternalId

        Returns
        -------
        int
        """
        item = self._idToItem.get(id)
        if item is None:
            raise Tf.ErrorException("Cannot find '%d' in PrimIdTable" % id)
        return len(item.children)

    def GetChildPath(self, id, index):
        """
        Parameters
        ----------
        id : _InternalId
        index : int

        Returns
        -------
        Sdf.Path
        """
        item = self._idToItem.get(id)
        if item is None:
            raise Tf.ErrorException("Cannot find '%d' in PrimIdTable" % id)

        if index >= len(item.children):
            raise Tf.ErrorException(
                "Index '%d' exceeds number of children of '%s' in PrimIdTable" %
                (index, item.path))

        return item.children[index]

    def GetParentId(self, id):
        """
        Parameters
        ----------
        id : _InternalId

        Returns
        -------
        _InternalId
        """
        item = self._idToItem.get(id)
        if item is None:
            raise Tf.ErrorException("Cannot find '%d' in PrimIdTable" % id)

        path = item.path
        if path == self._root:
            return 0

        parent = path.GetParentPath()
        parentId = self._pathToId.get(parent)
        if parentId is None:
            raise Tf.ErrorException(
                "Cannot find parent '%s' of '%s' in PrimIdTable" %
                (parent, path))

        return parentId

    def GetLastId(self):
        """
        Returns
        -------
        _InternalId
        """
        return self._nextAvailableId - 1

    def GetRow(self, id):
        """
        Parameters
        ----------
        id : _InternalId

        Returns
        -------
        int
        """
        item = self._idToItem.get(id)
        if item is None:
            raise Tf.ErrorException("Cannot find '%d' in PrimIdTable" % id)

        parentId = self.GetParentId(id)
        path = self.GetPathFromId(id)
        if path == self._root:
            return 0

        parentItem = self._idToItem.get(parentId)
        if parentItem is None:
            raise Tf.ErrorException(
                "Cannot find parent '%d' of '%d' in PrimIdTable" %
                (parentId, id))

        return parentItem.index(path)
        # # Run a search to find the index.
        # rowIterator = std.find(parentItem.children.begin(),
        #                        parentItem.children.end(), path)
        #
        # return rowIterator - parentItem.children.begin()

    def RegisterChild(self, id, index):
        """
        Parameters
        ----------
        id : _InternalId
        index : int

        Returns
        -------
        bool
        """
        item = self._idToItem.get(id)
        if item is None:
            raise Tf.ErrorException("Cannot find '%d' in PrimIdTable" % id)

        if index >= len(item.children):
            raise Tf.ErrorException(
                "Index '%d' exceeds number of children of '%s' in PrimIdTable" %
                (index, item.path))

        path = item.children[index]

        if path in self._pathToId:
            # already registered.
            return True

        prim = self._stage.GetPrimAtPath(path)
        if not prim:
            raise Tf.ErrorException(
                "Expected child (%d) of id (%d) has expired.", (index, id))

        if self._nextAvailableId >= self._maxId:
            _logger.warn("out of indices.")
            return False

        self._RegisterPrim(prim)
        return True

    def _DeleteSubtree(self, path):
        """
        Parameters
        ----------
        path : Sdf.Path

        Returns
        -------
        None
        """
        if path in self._pathToId:
            _logger.debug("Deleting instantiated path: '%s'", path)
            id = self._pathToId.pop(path)
            for childPath in self._idToItem[id].children:
                self._DeleteSubtree(childPath)
            self._idToItem.pop(id)
        else:
            _logger.debug(
                "Skipping deletion of uninstantiated path: '%s'", path)

    def _InvalidateSubtree(self, path):
        """
        Parameters
        ----------
        path : Sdf.Path

        Returns
        -------
        None
        """
        id = self._pathToId.get(path)
        if id is not None:
            prim = self._stage.GetPrimAtPath(path)
            if prim.IsValid() and self._predicate(prim):
                _logger.debug("Keeping '%s' during invalidation.", path)
                for childPath in self._idToItem[id].children:
                    self._InvalidateSubtree(childPath)

                _logger.debug("Original size: %d children.",
                              len(self._idToItem[id].children))
                self._idToItem[id].children[:] = []
                for child in prim.GetFilteredChildren(self._predicate):
                    self._idToItem[id].children.append(child.GetPath())

                _logger.debug("New size: %d children.\n",
                              len(self._idToItem[id].children))
            else:
                _logger.debug("Rejecting '%s' during invalidation.", path)
                self._DeleteSubtree(path)

        else:
            _logger.debug("Skipping invalidation of uninstantiated path '%s'",
                          path)

    def ResyncSubtrees(self, paths):
        """
        Parameters
        ----------
        paths : List[Sdf.Path]

        Returns
        -------
        None
        """
        # TODO: Is there a way to optimize the paths that need to be traversed
        # if there are redundancies? (say, /World/foo and /World/foo/bar)
        sortedPaths = paths[:]
        sortedPaths.sort()

        # Uniquify the list of parents.
        uniqueParents = set()  # type: Set[Sdf.Path]
        for path in sortedPaths:
            uniqueParents.add(path.GetParentPath())

        outOfSyncPaths = set()  # type: Set[Sdf.Path]

        # Update the list of children per unique parent.
        for parentPath in uniqueParents:
            _logger.debug("Updating children of parent: '%s'\n", parentPath)

            parentId = self.GetIdFromPath(parentPath)
            sortedOriginalChildren = self._idToItem[parentId].children
            sortedOriginalChildren.sort()

            newChildren = []  # type: List[Sdf.Path]

            # Look through the new children to find any paths aren't in the
            # original
            # children and not in sorted paths.  These have gotten out of sync.
            parentPrim = self._stage.GetPrimAtPath(parentPath)
            for child in parentPrim.GetFilteredChildren(self._predicate):
                inOriginalChildren = binarySearch(sortedOriginalChildren,
                                                  child.GetPath())
                inResyncPaths = binarySearch(sortedPaths, child.GetPath())

                if inOriginalChildren or inResyncPaths:
                    _logger.debug("Keeping child: '%s'", child.GetPath())
                    newChildren.append(child.GetPath())
                else:
                    _logger.debug(
                        "Out of sync new child: '%s'", child.GetPath())
                    outOfSyncPaths.add(child.GetPath())

            sortedNewChildren = newChildren[:]
            # std.sort(sortedNewChildren.begin(), sortedNewChildren.end())
            sortedNewChildren.sort()
            # Look through the original children to find any paths are missing
            # and not in sorted paths.  These have gotten out of sync, likely
            # because ResyncSubtrees has been called with an incomplete list.
            # This isn't strictly necessary other than for error checking.
            for childPath in sortedOriginalChildren:
                inNewChildren = binarySearch(sortedNewChildren, childPath)
                inResyncPaths = binarySearch(sortedPaths, childPath)
                if not inNewChildren and not inResyncPaths:
                    _logger.debug(
                        "Out of sync original child: '%s'", childPath)
                    outOfSyncPaths.add(childPath)

            # Assign the new children vector to the parent.
            self._idToItem[parentId].children[:] = newChildren
            _logger.debug("Total children count: '%d'",
                          len(self._idToItem[parentId].children))

        if outOfSyncPaths:
            raise Tf.ErrorException("Indices may have been lost during "
                                    "index resync.")
        for path in sortedPaths:
            self._InvalidateSubtree(path)

    def PrintFullIndex(self):
        """
        Returns
        -------
        None
        """
        print("Root: %s " % self._root)
        for id, item in self._idToItem:
            if item.path in self._pathToId:
                if self._pathToId[item.path] == id:
                    pathMapInfo = "correct"
                else:
                    pathMapInfo = "out of sync path map entry: %s" % \
                                  self._pathToId[item.path]
            else:
                pathMapInfo = "Missing Path Map Entry."
            print("[id]: %d [path]: %s [path map]: %s [child count]: %d" %
                  (id, item.path, pathMapInfo, len(item.children)))

        for path, id in self._pathToId.items():
            if id not in self._idToItem:
                print("Dangling path: %s" % path)

    def PrintSubtreeIndex(self, path):
        """
        Parameters
        ----------
        path : Sdf.Path

        Returns
        -------
        None
        """
        if path not in self._pathToId:
            print("(uninstantiated) %s" % path)
        else:
            id = self.GetIdFromPath(path)
            print("%s id: %d row: %d" % (path, id, self.GetRow(id)))
            for child in self._idToItem[id].children:
                self.PrintSubtreeIndex(child)
